src/Draco.Compiler/Internal/Declarations/DeclarationTable.cs

```diff
@@ -15,46 +19,99 @@ internal sealed class DeclarationTable
+
+    private readonly Compilation compilation;
     private readonly ImmutableArray<SyntaxTree> syntaxTrees;
 
-    private DeclarationTable(ImmutableArray<SyntaxTree> syntaxTrees)
+    private DeclarationTable(ImmutableArray<SyntaxTree> syntaxTrees, Compilation compilation)
     {
         this.syntaxTrees = syntaxTrees;
+        this.compilation = compilation;
     }
 
-    // NOTE: We don't have modules specified yet, so all added syntaxes are assumed to be in a global module with empty name
-    private MergedModuleDeclaration BuildMergedRoot() =>
-        new(this.syntaxTrees.Select(s => new SingleModuleDeclaration(string.Empty, (CompilationUnitSyntax)s.Root)).ToImmutableArray());
+    private MergedModuleDeclaration BuildMergedRoot()
+    {
+        // If we don't have root path, we put all file into top level module
+        if (string.IsNullOrEmpty(this.RootPath))
+        {
+            var singleModules = this.syntaxTrees
+                .Select(s => new SingleModuleDeclaration(string.Empty, string.Empty, (CompilationUnitSyntax)s.Root))
+                .ToImmutableArray();
+            return new(string.Empty, string.Empty, singleModules);
+        }
+
+        var rootName = Path.GetFileName(this.RootPath.TrimEnd(Path.DirectorySeparatorChar));
+        var modules = ImmutableArray.CreateBuilder<SingleModuleDeclaration>();
+        foreach (var tree in this.syntaxTrees)
+        {
+            var path = Path.TrimEndingDirectorySeparator(tree.SourceText.Path?.LocalPath ?? string.Empty);
+
+            // In memory tree, default to root module
+            if (string.IsNullOrEmpty(path))
+            {
+                modules.Add(new SingleModuleDeclaration(rootName, rootName, (CompilationUnitSyntax)tree.Root));
+                continue;
+            }
+
+            // Add error if path doesn't start with root path
+            if (!path.StartsWith(this.RootPath))
+            {
+                this.compilation.GlobalDiagnosticBag.Add(
+                Diagnostic.Create(
+                    template: SymbolResolutionErrors.FilePathOutsideOfRootPath,
+                    location: null,
+                    path, this.RootPath));
+
+                // Add to root so the compilation can continue
+                modules.Add(new SingleModuleDeclaration(rootName, rootName, (CompilationUnitSyntax)tree.Root));
+                continue;
+            }
+
+            var subPath = path[this.RootPath.Length..].TrimStart(Path.DirectorySeparatorChar);
+            var fullName = Path.TrimEndingDirectorySeparator(Path.GetDirectoryName(subPath) ?? string.Empty).Replace(Path.DirectorySeparatorChar, '.');
+
+            // Root module
+            if (string.IsNullOrEmpty(fullName)) fullName = rootName;
+            else fullName = $"{rootName}.{fullName}";
+            modules.Add(new SingleModuleDeclaration(fullName.Split('.').Last(), fullName, (CompilationUnitSyntax)tree.Root));
```

Comment:
Suggestion, factor out the name into a variable assignment to have a self-documenting name. Also don't use `.Last()` on an array, use `[^1]`